lib/spandex_datadog/sampling_strategies/use_agent_sampling_rate.ex

```diff
@@ -0,0 +1,49 @@
+defmodule SpandexDatadog.SamplingStrategies.UseAgentSamplingRate do
+  @moduledoc """
+  This sampling strategy uses the sampling rate set in the Datadog Agent.
+
+  The datadog agent responds to sending it traces with a sampling rate calculated
+  Using an algorithm that takes into account it's configuration, rate limiting, service traffic etc.
+  The tracing library is then able to use these rates to set the priority of the trace.
+
+  We keep the current advised sampling rate in the process dictionary and use that to calculate the priority.
+  """
+
+  @behaviour Spandex.SamplingStrategy
+
+  @keep_all_traces 1
+
+  @max_uint64 18_446_744_073_709_551_615
+  @knuth_factor 111_111_111_111_111_1111
+
+  @impl true
+  def calculate_priority(trace_id, opts \\ []) do
+    sample_rate = get_sampling_rate(opts)
+
+    threshold = trunc(sample_rate * @max_uint64)
+
+    if rem(trace_id * @knuth_factor, @max_uint64) <= threshold,
+      do: 1,
+      else: 0
```

Comment:
Is there any documentation on why is this calculated the way it is? I remember this from `ddtrace` but there was no explanation.

Also this calculation seems to be generic for any rate-based sampling strategy in `ddtrace` - which reminds me that you're not implementing a generic rate sampler that would take the sampling rate from service config instead of the Agent.

Reply:
from what I understand it's basically a super simple way to make the trace_ids map to evenly distributed numbers? a super-simple "uniform-enough" hashing.

And as for generic rate sampling. I just saw it as out of scope for this PR. It was not the goal for now to replicate all the features of the official libraries with regards to sampling. For us the agent-based sampling strategy I think is enough, along with introducing the abstraction of a sampling_strategy so anyone can make their own if they need to